pyro/contrib/gp/kernels/coregionalize.py

```diff
@@ -0,0 +1,57 @@
+from __future__ import absolute_import, division, print_function
+
+import torch
+from torch.nn import Parameter
+
+from .kernel import Kernel
+
+
+class Coregionalize(Kernel):
+    r"""
+    A kernel for the linear model of coregionalization
+    :math:`k(x,z) = x R R^T z^T` where :math:`R` is an
+    ``input_dim``-by-``rank`` regionalization matrix and typically
+    ``rank < input_dim``.
+
+    This generalizes the
+    :class:`~pyro.contrib.gp.kernels.dot_product.Linear` kernel to multiple
+    features.
+
+    References:
+
+    [1] Mauricio A. Alvarez, Lorenzo Rosasco, Neil D. Lawrence (2012)
+        Kernels for Vector-Valued Functions: a Review
+
+    :param int input_dim: Number of feature dimensions of inputs.
+    :param torch.Tensor regions: An optional ``(input_dim, rank)`` shaped
+        matrix that maps features to ``rank``-many regions. If unspecified,
+        this will be randomly initialized.
```

Comment:
Should we check the positive semidefinite of this kernel? To guarantee it, we can add an additional diagonal term (the form is similar to LowRankMultivariateNormal).

Reply:
Sure, I can add a vector of independent variances to make this numerically stable (note that mathematically `R R^t` will always be positive semidefinite). In an effort to avoid greek letters, could rename `regions` to `components` (in analogy with principle components analysis), and denote this vector of independent noises as `bias` (in analogy with the `Polynomial` kernel)?

Reply:
Actually I don't understand the motivation for the `kappa` part of the kernel and it seems to complicate the implementation. Do you understand the motivation, or why we should bake a vector of variances into this implementation rather than separate that out as a distinct `Kernel` class?

Reply:
Yeah, agree! The `components` name seems better to me. I also think that `kappa` is unnecessary and only used to guarantee the positiveness of coregionalization matrix (in case of R has rank less than `r`). I guess ignoring it will not be an issue (hopefully) because we already have `jitter` term in GP models to deal with the singularity of covariance matrix.

Reply:
@fehiepsi I've noticed that this kernel is sensitive to initialization. I've implemented the `kappa` thing locally to experiment. If it appears to make things more robust (more convex), I'll send a follow-up PR. Thanks for suggesting this!

Reply:
Here's the follow-up PR: https://github.com/uber/pyro/pull/1116